### PR TITLE
HHH-16050 Optimize iteration efficiency in StandardStack 

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-16050

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java [hunk 1]
```diff
@@ -6,139 +6,159 @@
...
+	private int head = 0;
+	private int tail = 0;
 
-	private ArrayDeque<T> internalStack;
-	private static final Object NULL_TOKEN = new Object();
-
-	public StandardStack() {
+	@SuppressWarnings("unchecked")
+	public StandardStack(Class<T> type) {
+		elements = (T[]) Array.newInstance( type, 8 );
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java:36]: There's a non-negligible change that the `StandardStack` might be constructed, but then nothing gets added.

So I think it would be best to allocate the Array lazily?

File: hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java [hunk 2]
```diff
@@ -6,139 +6,159 @@
...
-
-	private Deque<T> stackInstanceExpected() {
-		if ( internalStack == null ) {
-			//"7" picked to use 8, but skipping the odd initialCapacity method
-			internalStack = new ArrayDeque<>( 7 );
+	public void push(T e) {
+		head = dec( head );
+		elements[head] = e;
+		if ( head == tail ) {
+			grow();
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java:49]: It's a bit surprising that we might end up growing the array right *after* having stored the last element.

Would be fantastic to avoid growing after a push, but rather make space if there is no space before a push?

I guess this also relates with the reason about why you needed to pre-allocate the array even on construction. Neither should be necessary.

File: hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java [hunk 3]
```diff
@@ -6,139 +6,159 @@
...
 	@Override
 	public void clear() {
-		if ( internalStack != null ) {
-			internalStack.clear();
+		circularClear( elements, head, tail );
+		head = tail = 0;
+	}
+
+	private static void circularClear(Object[] es, int i, int end) {
+		for ( int to = ( i <= end ) ? end : es.length; ; i = 0, to = end ) {
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java:95]: I realize now that I gave you a bad suggestion, sorry.
In our particular case we don't really need a circular array; I had forgotten that `ArrayDeque` is based on a circular array, I guess because it allows adding elements either at the end or at the beginning of the queue (possibly needing to shift all entries).

but our use case is much simpler: we only have pop/push modification operations; this implies you should be able to simplify this all a little bit, as you don't need to keep track of both the head and the tail, only needing to ensure the array grows if there is not enough space: we always only insert new elements at the end.

The good news: you can indeed make this much more efficient than the existing implementation :)

File: hibernate-core/src/test/java/org/hibernate/orm/test/util/StandardStackTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	@Test
+	public void testClear() {
+		final Stack<Integer> stack = allocateStack( 42 );
+		assertEquals( 42, stack.depth() );
+		assertFalse( stack.isEmpty() );
+		stack.clear();
+		assertEquals( 0, stack.depth() );
+		assertTrue( stack.isEmpty() );
```
Comment (Sanne) [inline hibernate-core/src/test/java/org/hibernate/orm/test/util/StandardStackTest.java:43]: let's also test for that the various "find" methods work correctly on both an empty and a cleared stack?

**Review by Sanne — CHANGES_REQUESTED** [review]
Excellent work!

But I'm suggesting a small change in the algorythm as I just realized we can do even better than the original ArrayDeque in our context :)

No rush - and great job!

File: hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java (hunk 1 above)
Comment (Sanne) [reply hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java:36]: N.B. when I wrote this comment I hadn't realized the implementation was based on a circular array. This one might get clearer if you read the other comments first.

File: hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java (hunk 3 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/internal/util/collections/StandardStack.java:95]: You're right, I was too focused on reproducing the previous behavior that I didn't stop and think if there was a better way of achieving this. The implementation will also be a lot easier tracking just the top of the stack. Thanks for all the suggestions!

**Review by Sanne — APPROVED** [review]
Thanks, looks great!
